cranelift/interpreter/src/step.rs

```diff
@@ -1403,35 +1426,41 @@ pub enum CraneliftTrap {
 }
 
 /// Compare two values using the given integer condition `code`.
-fn icmp<V>(ctrl_ty: types::Type, code: IntCC, left: &V, right: &V) -> ValueResult<V>
-where
-    V: Value,
-{
-    let cmp = |bool_ty: types::Type, code: IntCC, left: &V, right: &V| -> ValueResult<V> {
-        Ok(Value::bool(
+fn icmp(
+    ctrl_ty: types::Type,
+    code: IntCC,
+    left: &DataValue,
+    right: &DataValue,
+) -> ValueResult<DataValue> {
+    let cmp = |bool_ty: types::Type,
+               code: IntCC,
+               left: &DataValue,
+               right: &DataValue|
+     -> ValueResult<DataValue> {
+        Ok(DataValueExt::bool(
             match code {
-                IntCC::Equal => Value::eq(left, right)?,
-                IntCC::NotEqual => !Value::eq(left, right)?,
-                IntCC::SignedGreaterThan => Value::gt(left, right)?,
-                IntCC::SignedGreaterThanOrEqual => Value::ge(left, right)?,
-                IntCC::SignedLessThan => Value::lt(left, right)?,
-                IntCC::SignedLessThanOrEqual => Value::le(left, right)?,
-                IntCC::UnsignedGreaterThan => Value::gt(
-                    &left.clone().convert(ValueConversionKind::ToUnsigned)?,
-                    &right.clone().convert(ValueConversionKind::ToUnsigned)?,
-                )?,
-                IntCC::UnsignedGreaterThanOrEqual => Value::ge(
-                    &left.clone().convert(ValueConversionKind::ToUnsigned)?,
-                    &right.clone().convert(ValueConversionKind::ToUnsigned)?,
-                )?,
-                IntCC::UnsignedLessThan => Value::lt(
-                    &left.clone().convert(ValueConversionKind::ToUnsigned)?,
-                    &right.clone().convert(ValueConversionKind::ToUnsigned)?,
-                )?,
-                IntCC::UnsignedLessThanOrEqual => Value::le(
-                    &left.clone().convert(ValueConversionKind::ToUnsigned)?,
-                    &right.clone().convert(ValueConversionKind::ToUnsigned)?,
-                )?,
+                IntCC::Equal => left == right,
+                IntCC::NotEqual => !(left == right),
```

Comment:
This can be just `!=`, can't it?
```suggestion
                IntCC::NotEqual => left != right,
```

Reply:
Right, I was considering that for floats it isn't equal, but thise can't occur here.